core/arch/arm/mm/core_mmu.c

```diff
@@ -480,3 +492,222 @@ unsigned int cache_maintenance_l2(int op, paddr_t pa, size_t len)
+{
+	uint32_t exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	bool ret = false;
+
+#ifdef ARM32
+#ifdef CFG_WITH_LPAE
+	uint64_t par;
+
+	write_ats1cpw((uint32_t)va);
+	isb();
+	par = read_par64();
+	if (par & PAR_F)
+		goto out;
+	*pa = (par & (PAR64_PA_MASK << PAR_PA_SHIFT)) |
+		((uint32_t)va & ((1 << PAR_PA_SHIFT) - 1));
+#else
+	uint32_t par;
+
+	write_ats1cpw((uint32_t)va);
+	isb();
+	par = read_par32();
+	if (par & PAR_F)
+		goto out;
+	*pa = (par & (PAR32_PA_MASK << PAR_PA_SHIFT)) |
+		((uint32_t)va & ((1 << PAR_PA_SHIFT) - 1));
+#endif
+#endif /*ARM32*/
+
+#ifdef ARM64
+	uint64_t par;
+
+	write_at_s1e1r((uint64_t)va);
+	isb();
+	par = read_par_el1();
+	if (par & PAR_F)
+		goto out;
+	*pa = (par & (PAR_PA_MASK << PAR_PA_SHIFT)) |
+		((uint64_t)va & ((1 << PAR_PA_SHIFT) - 1));
+#endif
+
+	ret = true;
+out:
+	thread_unmask_exceptions(exceptions);
+	return ret;
+}
+
+#if defined(CFG_TEE_CORE_DEBUG) && CFG_TEE_CORE_DEBUG != 0
+static void check_virt_to_phys(void *va, paddr_t pa)
+{
+	TEE_Result res;
+	vaddr_t user_va_base;
+	size_t user_va_size;
+	vaddr_t v = (vaddr_t)va;
+	paddr_t p = 0;
+
+	core_mmu_get_user_va_range(&user_va_base, &user_va_size);
+	if (v >= user_va_base && v < (user_va_base + user_va_size)) {
+		if (!core_mmu_user_mapping_is_active()) {
+			TEE_ASSERT(pa == 0);
+			return;
+		}
+
+		res = tee_mmu_user_va2pa_helper(
+			to_user_ta_ctx(tee_mmu_get_ctx()), va, &p);
+		if (res == TEE_SUCCESS)
+			TEE_ASSERT(pa == p);
+		else
+			TEE_ASSERT(pa == 0);
+		return;
+	}
+#ifdef CFG_WITH_PAGER
+	if (v >= (CFG_TEE_LOAD_ADDR & ~CORE_MMU_PGDIR_MASK) &&
+	    v <= (CFG_TEE_LOAD_ADDR | CORE_MMU_PGDIR_MASK)) {
+		struct core_mmu_table_info *ti = &tee_pager_tbl_info;
+		uint32_t a;
+
+		/*
+		 * Lookups in the page table managed by the pager is
+		 * dangerous for addresses in the paged area as those pages
+		 * changes all the time. But some ranges are safe, rw areas
+		 * when the page is populated for instance.
+		 */
+		core_mmu_get_entry(ti, core_mmu_va2idx(ti, v), &p, &a);
+		if (a & TEE_MATTR_VALID_BLOCK) {
+			paddr_t mask = ((1 << ti->shift) - 1);
+
+			p |= v & mask;
+			TEE_ASSERT(pa == p);
+		} else
+			TEE_ASSERT(pa == 0);
+		return;
+	}
+#endif
+	if (v >= TEE_MMU_KMAP_START_VA && v < TEE_MMU_KMAP_END_VA) {
+		void *void_pa;
+
+		res = tee_mmu_kmap_va2pa_helper(va, &void_pa);
+		if (res == TEE_SUCCESS)
+			TEE_ASSERT(pa == (paddr_t)(uintptr_t)void_pa);
+		else
+			TEE_ASSERT(pa == 0);
+		return;
+	}
+	if (!core_va2pa_helper(va, &p))
+		TEE_ASSERT(pa == p);
+	else
+		TEE_ASSERT(pa == 0);
+}
+#else
+static void check_virt_to_phys(void *va __unused, paddr_t pa __unused)
+{
+}
+#endif
+
+paddr_t virt_to_phys(void *va)
+{
+	paddr_t pa;
+
+	if (!arm_va2pa_helper(va, &pa))
+		pa = 0;
+	check_virt_to_phys(va, pa);
+	return pa;
+}
+
+#if defined(CFG_TEE_CORE_DEBUG) && CFG_TEE_CORE_DEBUG != 0
+static void check_phys_to_virt(paddr_t pa, void *va)
+{
+	TEE_ASSERT(!va || virt_to_phys(va) == pa);
```

Comment:
`TEE_ASSERT((!va && !pa) || virt_to_phys(va) == pa);` instead as `map_pa2va()` return `NULL` in case `!map`. So we'll get `va=NULL`, whereas it looks an error to me in such a case